From d598ae4cf36fdb46326bdebe2c7d7957c88a7827 Mon Sep 17 00:00:00 2001
From: Radharapu Rakesh <rakesh.radharapu@amd.com>
Date: Mon, 10 Feb 2025 18:58:09 -1200
Subject: [PATCH 099/110] net: xilinx: axienet: Fix NULL Pointer dereference up
 on skb corruption

GRO Skb data is corrupted when multiple ethernet interfaces are running
with GRO in parallel using the same queue number.

Remove global skb definition and add the information per queue to avoid
this error and ensure proper functionality on multi-ethernet subsystems.

Signed-off-by: Radharapu Rakesh <rakesh.radharapu@amd.com>
Reviewed-by: Harini Katakam <harini.katakam@amd.com>
Message-ID: <20250211065809.1303935-1-rakesh.radharapu@amd.com>
---
 drivers/net/ethernet/xilinx/xilinx_axienet.h  |  4 ++
 .../net/ethernet/xilinx/xilinx_axienet_eoe.c  | 38 +++++++++----------
 2 files changed, 22 insertions(+), 20 deletions(-)

diff --git a/drivers/net/ethernet/xilinx/xilinx_axienet.h b/drivers/net/ethernet/xilinx/xilinx_axienet.h
index e956ca882cd78..12b31a4e39f05 100644
--- a/drivers/net/ethernet/xilinx/xilinx_axienet.h
+++ b/drivers/net/ethernet/xilinx/xilinx_axienet.h
@@ -1155,6 +1155,8 @@ struct axienet_local {
  * @txq_bytes:   Number of transmit bytes processed by the dma queue.
  * @rxq_packets: Number of receive packets processed by the dma queue.
  * @rxq_bytes:	Number of receive bytes processed by the dma queue.
+ * @skb:	Socket buffer for GRO.
+ * @rx_data:	stores the length of GRO skb fragments.
  */
 struct axienet_dma_q {
 	struct axienet_local	*lp; /* parent */
@@ -1193,6 +1195,8 @@ struct axienet_dma_q {
 	unsigned long txq_bytes;
 	unsigned long rxq_packets;
 	unsigned long rxq_bytes;
+	struct sk_buff *skb;
+	u32 rx_data;
 };
 
 #define AXIENET_ETHTOOLS_SSTATS_LEN 6
diff --git a/drivers/net/ethernet/xilinx/xilinx_axienet_eoe.c b/drivers/net/ethernet/xilinx/xilinx_axienet_eoe.c
index cf9ad494af4e2..31577fd6f9fdd 100644
--- a/drivers/net/ethernet/xilinx/xilinx_axienet_eoe.c
+++ b/drivers/net/ethernet/xilinx/xilinx_axienet_eoe.c
@@ -224,8 +224,6 @@ int axienet_eoe_recv_gro(struct net_device *ndev, int budget,
 			 struct axienet_dma_q *q)
 {
 	struct axienet_local *lp = netdev_priv(ndev);
-	static struct sk_buff *skb[XAE_MAX_QUEUES];
-	static u32 rx_data[XAE_MAX_QUEUES];
 	u32 length, packets = 0, size = 0;
 	unsigned int numbdfree = 0;
 	struct aximcdma_bd *cur_p;
@@ -255,39 +253,39 @@ int axienet_eoe_recv_gro(struct net_device *ndev, int budget,
 
 		page_addr = page_address(page);
 
-		rx_data[q->chan_id - 1] += length;
+		q->rx_data += length;
 
-		if (skb[q->chan_id - 1]) {
-			skb_add_rx_frag(skb[q->chan_id - 1],
-					skb_shinfo(skb[q->chan_id - 1])->nr_frags,
-					page, 0, length, rx_data[q->chan_id - 1]);
+		if (q->skb) {
+			skb_add_rx_frag(q->skb,
+					skb_shinfo(q->skb)->nr_frags,
+					page, 0, length, q->rx_data);
 		}
 
 		if (((cur_p->app0 & XEOE_UDP_GRO_RXSOP_MASK) >> XEOE_UDP_GRO_RXSOP_SHIFT)) {
 			/* Allocate new skb and update in BD */
-			skb[q->chan_id - 1] = netdev_alloc_skb(ndev, length);
-			memcpy(skb[q->chan_id - 1]->data, page_addr, length);
-			skb_put(skb[q->chan_id - 1], length);
+			q->skb = netdev_alloc_skb(ndev, length);
+			memcpy(q->skb->data, page_addr, length);
+			skb_put(q->skb, length);
 			put_page(page);
 		} else if (((cur_p->app0 & XEOE_UDP_GRO_RXEOP_MASK) >> XEOE_UDP_GRO_RXEOP_SHIFT)) {
-			skb_set_network_header(skb[q->chan_id - 1], XEOE_MAC_HEADER_LENGTH);
-			iphdr = (struct iphdr *)skb_network_header(skb[q->chan_id - 1]);
-			skb_set_transport_header(skb[q->chan_id - 1],
+			skb_set_network_header(q->skb, XEOE_MAC_HEADER_LENGTH);
+			iphdr = (struct iphdr *)skb_network_header(q->skb);
+			skb_set_transport_header(q->skb,
 						 iphdr->ihl * 4 + XEOE_MAC_HEADER_LENGTH);
-			uh = (struct udphdr *)skb_transport_header(skb[q->chan_id - 1]);
+			uh = (struct udphdr *)skb_transport_header(q->skb);
 
 			/* App Fields are in Little Endian Byte Order */
 			iphdr->tot_len = htons(cur_p->app1 & XEOE_UDP_GRO_PKT_LEN_MASK);
 			iphdr->check = (__force __sum16)htons((cur_p->app1 &
 					XEOE_UDP_GRO_RX_CSUM_MASK) >> XEOE_UDP_GRO_RX_CSUM_SHIFT);
 			uh->len = htons((cur_p->app1 & XEOE_UDP_GRO_PKT_LEN_MASK) - iphdr->ihl * 4);
-			skb[q->chan_id - 1]->protocol = eth_type_trans(skb[q->chan_id - 1], ndev);
-			skb[q->chan_id - 1]->ip_summed = CHECKSUM_UNNECESSARY;
-			rx_data[q->chan_id - 1] = 0;
+			q->skb->protocol = eth_type_trans(q->skb, ndev);
+			q->skb->ip_summed = CHECKSUM_UNNECESSARY;
+			q->rx_data = 0;
 			/* This will give SKB to n/w Stack */
-			if (skb_shinfo(skb[q->chan_id - 1])->nr_frags <= XEOE_UDP_GRO_MAX_FRAG) {
-				netif_receive_skb(skb[q->chan_id - 1]);
-				skb[q->chan_id - 1] = NULL;
+			if (skb_shinfo(q->skb)->nr_frags <= XEOE_UDP_GRO_MAX_FRAG) {
+				netif_receive_skb(q->skb);
+				q->skb = NULL;
 			}
 		}
 
-- 
2.43.0

